tests/integration/update.bats

```diff
@@ -584,10 +584,52 @@ EOF
 	check_cgroup_value "cpu.rt_runtime_us" 600001
 }
 
-@test "update devices [minimal transition rules]" {
+@test "update devices limits access [minimal transition rules]" {
 	[[ "$ROOTLESS" -ne 0 ]] && requires rootless_cgroup
-	# This test currently only makes sense on cgroupv1.
-	requires cgroups_v1
+
+	requires root
+
+	# Run a basic shell script that tries to read from to /dev/private,
+	# a dummy device it should not have access to. During "runc update",
+	# it should still not be able to read from the file/device.
+	update_config '.process.args |= ["sh", "-c", "while true; do head -c 10 /dev/private 2>&1 | grep -v \"head: /dev/private: Operation not permitted\"; done"]'
+	update_config '.linux.devices = [{"path": "/dev/private", "type": "c", "fileMode": 438, "major": 1, "minor": 100, "uid": 0, "gid": 0}]'
+	update_config '.linux.resources.devices |= [{"allow": false, "major": 1, "minor": 100, "access": "rwm"}]'
+
+	# Set up a temporary console socket and recvtty so we can get the stdio.
+	TMP_RECVTTY_DIR="$(mktemp -d "$BATS_RUN_TMPDIR/runc-tmp-recvtty.XXXXXX")"
+	TMP_RECVTTY_PID="$TMP_RECVTTY_DIR/recvtty.pid"
+	TMP_CONSOLE_SOCKET="$TMP_RECVTTY_DIR/console.sock"
+	CONTAINER_OUTPUT="$TMP_RECVTTY_DIR/output"
+	("$RECVTTY" --no-stdin --pid-file "$TMP_RECVTTY_PID" \
+		--mode single "$TMP_CONSOLE_SOCKET" &>"$CONTAINER_OUTPUT") &
+	retry 10 0.1 [ -e "$TMP_CONSOLE_SOCKET" ]
+
+	# Run the container in the background.
+	runc run -d --console-socket "$TMP_CONSOLE_SOCKET" test_update
+	cat "$CONTAINER_OUTPUT"
+	[ "$status" -eq 0 ]
+
+	# Trigger an update. This update doesn't actually change the device rules,
+	# but it will trigger the devices cgroup code to reapply the current rules.
+	# We trigger the update a few times to make sure we hit the race.
+	for _ in {1..30}; do
+		# TODO: Update "runc update" so we can change the device rules.
+		runc update --pids-limit 30 test_update
+		[ "$status" -eq 0 ]
+	done
+
+	# Kill recvtty.
+	kill -9 "$(<"$TMP_RECVTTY_PID")"
+
+	# There should've been no output from the container.
+	cat "$CONTAINER_OUTPUT"
+	[ -z "$(<"$CONTAINER_OUTPUT")" ]
+}
+@test "update devices does not deny access [minimal transition rules]" {
```

Comment:
The test you're adding very much duplicates the existing one, and copies a lot of code around.

Perhaps it's better to just amend the existing test with the reading from a denied device case, to avoid code duplication.

Reply:
Or, if you'd like to keep them separate, rename `@test "update devices [minimal transition rules]" {` to a function, add some variable arguments, and have two test cases call it with different args.

Reply:
Agree! Squashed them into one now! :+1: Thanks!